client/services/FeedbackGeneratorService.js

```diff
@@ -53,8 +54,46 @@ tie.factory('FeedbackGeneratorService', [
       }
     };
 
+    var _getBuggyOutputTestFeedback = function (
+      failingTest, codeEvalResult) {
+      var hintIndex = 0;
+      var buggyMessages = failingTest.getMessages();
+      var feedback = FeedbackObjectFactory.create(false);
+      var lastSnapshot = (
+        TranscriptService.getTranscript().getPreviousSnapshot());
+      if (lastSnapshot !== null) {
+        // This section makes sure to provide a new hint
+        // if the student gets stuck on the same bug by checking
+        // that they've submitted new code with the same error.
+        var previousFeedback = lastSnapshot.getFeedback();
+        var previousHintIndex = previousFeedback.getHintIndex();
+        if (previousHintIndex !== null && 
+          previousHintIndex < buggyMessages.length) {
+          var previousMessages = previousFeedback.getParagraphs();
+          // This could cause a problem if two different buggy outputs
+          // have the exact same hint, but that shouldn't be allowed.
+          if (previousMessages[0].getContent() == 
+            buggyMessages[previousHintIndex]) {
+            var previousCode = (
+              lastSnapshot.getCodeEvalResult().getCode());
+            if (previousCode === codeEvalResult.getCode()) {
+              hintIndex = previousHintIndex;
+            }
+            hintIndex = previousHintIndex + 1 % buggyMessages.length;
```

Comment:
Shouldn't there be an 'else' here? Otherwise you're just overwriting the previous value in all cases.

Also, not sure I understand the significance of the "%"?